docs/src/internals.md

```diff
@@ -0,0 +1,256 @@
+`FormulaTerm` returns a tuple of the numeric forms for the left- (response) and
+right-hand (predictor) sides.
+
+```@repl 1
+model_cols(f, df)
+```
+
+Model cols can also take a single row from a table, as a `NamedTuple`:
+
+```@repl 1
+using Tables
+model_cols(f, first(Tables.rowtable(df)))
+```
+
+Any `AbstractTerm` can be passed to `model_cols` with a table and returns one or
+more numeric arrays:
+
+
+
+## Extending `@formula` syntax
+
+Package authors may want to create additional syntax to the `@formula` DSL so
+their users can conveniently specify particular kinds of models.  StatsModels.jl
+provides mechanisms for such extensions that do _not_ rely on compile time
+"macro magic", but on standard julian mechanisms of multiple dispatch.
+
+Extensions have three components:
+1. **Syntax**: the julia function which is given special meaning inside a formula.
+2. **Context**: the model type(s) where this extension applies
+3. **Behavior**: how tabular data bis transformed under this extension
+
+### Example
+
+As an example, we'll add syntax for specifying a [polynomial
+regression](https://en.wikipedia.org/wiki/Polynomial_regression) model, which
+fits a regression using polynomial basis functions of a continuous predictor.
+
+The first step is to specify the **syntax** we're going to use.  While it's
+possible to use an existing function, the best practice is to define a new
+function to make dispatch less ambiguous.
+
+```@example 1
+using StatsBase # hide
+# syntax: best practice to define a _new_ function
+poly(x, n) = x^n
+
+# type of model where syntax applies
+const POLY_CONTEXT = Any
+
+# struct for behavior
+struct PolyTerm <: AbstractTerm
+    term::ContinuousTerm
+    deg::Int
+end
+
+Base.show(io::IO, p::PolyTerm) = print(io, "poly($(p.term), $(p.deg))")
+
+function StatsModels.apply_schema(t::FunctionTerm{typeof(poly)}, sch, Mod::Type{<:POLY_CONTEXT})
+    term = apply_schema(t.args_parsed[1], sch, Mod)
+    isa(term, ContinuousTerm) ||
+        throw(ArgumentError("PolyTerm only works with continuous terms (got $term)"))
+    deg = t.args_parsed[2]
+    isa(deg, ConstantTerm) ||
+        throw(ArgumentError("PolyTerm degree must be a number (got $deg)"))
+    PolyTerm(term, deg.n)
+end
+
+function StatsModels.model_cols(p::PolyTerm, d::NamedTuple)
+    col = model_cols(p.term, d)
+    reduce(hcat, (col.^n for n in 1:p.deg))
+end
+
+StatsBase.coefnames(p::PolyTerm) = coefnames(p.term) .* "^" .* string.(1:p.deg)
+```
+
+Now, we can use `poly` in a formula:
+
+```@repl 1
+data = DataFrame(y = rand(4), a = rand(4), b = [1:4;])
+f = @formula(y ~ 1 + poly(b, 2) * a)
+f = apply_schema(f, schema(data))
+model_cols(f.rhs, data)
+coefnames(f.rhs)
+```
+
+It's also possible to _block_ interpretation of the `poly` syntax as special in
+certain contexts by adding additional (more specific) methods.  For instance, we
+could block for the default model context of `Nothing`:
```

Comment:
When is this default context used? It's not clear to me when this kind of definition would be useful, nor whether it's recommended to do that.

Reply:
It's used when calling `apply_schema` with two arguments (I'll add a note about this).  I'm not sure if _this_ case is useful, but the point was more to illustrate how you can have fine-grained control over the context that your syntax should apply by dispatching on the model type than to show a particular example that's actually something that would be useful in practice.

Reply:
OK. Then maybe better say/show that you can override the behavior for a specific model type?

Reply:
I now show this for `GLM.LinearModel` vs. `GLM.GeneralizedLinearModel`